stan/math/gpu/matrix_gpu.hpp

```diff
@@ -0,0 +1,173 @@
+   * An OpenCL buffer allocates the memory in the device that
+   * is provided by the context.
+   */
+  cl::Buffer oclBuffer_;
+  const int rows_;
+  const int cols_;
+
+ public:
+  int rows() const { return rows_; }
+
+  int cols() const { return cols_; }
+
+  int size() const { return rows_ * cols_; }
+
+  const cl::Buffer& buffer() const { return oclBuffer_; }
+
+  matrix_gpu() : rows_(0), cols_(0) {}
+
+  matrix_gpu(const matrix_gpu& a) : rows_(a.rows()), cols_(a.cols()) {
+    // this check is needed because creating OpenCL
+    // buffers of size 0 throws an OpenCL exception
+    if (a.size() > 0) {
+      // the queue is needed to enqueue the kernel for execution
+      cl::CommandQueue& cmdQueue = opencl_context.queue();
+      // the context is needed to create the buffer object
+      cl::Context& ctx = opencl_context.context();
+      try {
+        // creates a read&write object for "size" double values
+        // in the provided context
+        oclBuffer_
+            = cl::Buffer(ctx, CL_MEM_READ_WRITE, sizeof(double) * size());
+        /**
+         * Sets the arguments for the kernel. See copy_matrix_kernel in
+         * kernels/basic_matrix_gpu_kernels.hpp for the kernel code.
+         * The arguments are the source & destination matrices
+         * and the size in rows and columns.
+         */
+        // retrieves the kernel that copies memory from the
+        // input matrix a
+        cl::Kernel kernel = opencl_context.get_kernel("copy");
+        kernel.setArg(0, a.buffer());
+        kernel.setArg(1, buffer());
+        kernel.setArg(2, rows());
+        kernel.setArg(3, cols());
+        /**
+         * Runs the specified kernel with provided number of threads.
+         * - the first argument is the kernel object
+         * - the second argument is the thread offset that is used for
+         *   calculating the global thread ID, NULL here, meaning the
+         *   offset is 0,0
+         * - the third argument specifies the amount of threads to create
+         *   in 2D (rows, columns)
+         * - the fourth argument specifies the size of the thread block,
+         *   NULL here, meaning the OpenCL driver determines the size
+         * - the last two arguments are for tracking events associated
+         *   with the kernel (enqueue,start, stop,...) for profiling.
+         *   Not needed here.
+         *
+         */
+        cmdQueue.enqueueNDRangeKernel(kernel, cl::NullRange,
+                                      cl::NDRange(rows(), cols()),
+                                      cl::NullRange, NULL, NULL);
+      } catch (const cl::Error& e) {
+        check_opencl_error("copy GPU->GPU", e);
+      }
+    }
+  }
+  /**
+   * Constructor for the matrix_gpu that
+   * only allocates the buffer on the GPU.
+   *
+   * @param rows number of matrix rows
+   * @param cols number of matrix columns
+   *
+   * @throw <code>std::invalid_argument</code> if the
```

Comment:
This comment is not correct. Please update.

Reply:
Done.